Add bindings for using sockets

Duhemm opened the pull request.

muxanick commented:
Thanks!

rwhaling commented on nativelib/src/main/resources/posix/sys/socket.c:
```diff
@@ -0,0 +1,123 @@
+void scalanative_convert_sockaddr(struct scalanative_sockaddr *in,
+                                  struct sockaddr *out) {
+    out->sa_family = in->sa_family;
+    void *ignored = memcpy(out->sa_data, in->sa_data, 14);
```
I'm not sure there's a better way to do this for now, but I think this will clip off some data if sockaddr is actually a pointer to a larger struct like sockaddr_in6 or sockaddr_storage.

Duhemm replied:
[The posix spec for `struct sockaddr`](http://pubs.opengroup.org/onlinepubs/000095399/basedefs/sys/socket.h.html#tag_13_61) says that `sockaddr.sa_data` has a variable length. However, [its length has been `14` since the kernel uses git (at least)](https://github.com/torvalds/linux/blob/master/include/linux/socket.h#L29-L32).

[The length also appears to be `14` in Mac OS](https://github.com/opensource-apple/xnu/blob/10.12/bsd/sys/socket.h#L591-L595), though they say that it is actually larger. I'm not sure how to read that comment, but I've been using that code on both MacOS and Linux without any issue.

Indeed once we touch a `sockaddr_storage` or `sockaddr_in6` we'll lose some data. Another solution would be to define, in Scala Native, `sockaddr` as `CStruct0`, have it completely opaque to Scala code, and define functions in C to extract information from that `CStruct0`. I think that's the way to go.

Thanks for pointing out these issues!

densh replied:
I agree with @rwhaling, this approach doesn't seem to work here. Lets try `CStruct0` option.

asoltysik commented on nativelib/src/main/scala/scala/scalanative/posix/inttypes.scala:
```diff
@@ -0,0 +1,12 @@
+
+  type uint8_t  = CUnsignedChar
+  type uint16_t = CUnsignedShort
+  type uint32_t = CUnsignedInt
```
Is that true for a 64-bit processor?

densh replied:
type uint32_t = UInt

asoltysik commented:
@Duhemm getaddrinfo and getnameinfo seem to be missing.

Duhemm commented:
@asoltysik These two are defined in `<netdb.h>` which I didn't implement in this PR (I don't need it for the work on the test interface at the moment).

densh commented on nativelib/src/main/scala/scala/scalanative/posix/inttypes.scala:
```diff
@@ -0,0 +1,12 @@
+
+object inttypes {
+
+  type uint8_t  = CUnsignedChar
```
type uint8_t = UByte

densh commented on nativelib/src/main/scala/scala/scalanative/posix/inttypes.scala:
```diff
@@ -0,0 +1,12 @@
+object inttypes {
+
+  type uint8_t  = CUnsignedChar
+  type uint16_t = CUnsignedShort
```
type uint16_t = UShort

densh commented on nativelib/src/main/scala/scala/scalanative/posix/inttypes.scala:
```diff
@@ -0,0 +1,12 @@
+  type uint8_t  = CUnsignedChar
+  type uint16_t = CUnsignedShort
+  type uint32_t = CUnsignedInt
+
```
`type uint64_t = ULong`

asoltysik commented:
@Duhemm cool, I will do bindings for them then

Duhemm commented:
@densh I addressed your comments, this one should be ready for review. I tested it with the test interface on Linux and macOS, both work.

@rwhaling If you have additional comments, they would be greatly appreciated!

densh commented:
@Duhemm This PR needs some love from `scalafmt`.

Duhemm commented:
@densh Done

densh commented:
LGTM!
